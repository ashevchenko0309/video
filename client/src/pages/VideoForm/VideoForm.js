import React from "react"
import ReactRouterPropTypes from "react-router-prop-types"

import FORM_SCHEMA from "./video-from.schema"

import Container from "../../components/Grid/Container"
import Row from "../../components/Grid/Row"
import Column from "../../components/Grid/Column"
import InputField from "../../components/InputField/InputField"
import TextareaField from "../../components/TextareaField/TextareaField"
import VideoUploader from "../../components/VideoUploader/VideoUploader"
import FormErrorHint from "../../components/FormErrorHints/FormErrorHint"

import dataURLtoFile from "../../utils/dataUrlToFile"
import isValidLength from "../../utils/validations"

/* eslint-disable react/jsx-props-no-spreading */
class VideoFrom extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      formError: null,
      title: {
        hasError: null,
        errorMessage: "",
      },
      description: {
        hasError: null,
        errorMessage: "",
      },
      video: {
        hasError: null,
        errorMessage: "",
      },
    }
  }

  isFormValid = (title, description, videoFile) => {
    const titleValidationResult = isValidLength(
      title,
      FORM_SCHEMA.title.options.minLength,
      FORM_SCHEMA.title.options.maxLength
    )
    const descriptionValidationResult = isValidLength(
      description,
      FORM_SCHEMA.description.options.minLength,
      FORM_SCHEMA.description.options.maxLength
    )
    let isFormValid = true

    if (titleValidationResult.hasError) {
      this.setState({ title: { ...titleValidationResult } })
      isFormValid = false
    }

    if (descriptionValidationResult.hasError) {
      this.setState({ description: { ...descriptionValidationResult } })
      isFormValid = false
    }

    if (!videoFile) {
      this.setState({
        video: {
          hasError: true,
          errorMessage: "Must have video file",
        },
      })
      isFormValid = false
    }

    return isFormValid
  }

  onSubmit = (e) => {
    e.preventDefault()

    const { elements } = e.currentTarget

    const videoTitle = elements.title.value
    const videoDescription = elements.description.value
    const thumbUrlData = elements.thumb.value
    const videoFile = elements.video.files[0]

    if (!this.isFormValid(videoTitle, videoDescription, videoFile)) {
      this.setState({ formError: "Form have invalid value(s)" })
      return
    }

    const { formError } = this.state

    if (formError) {
      this.setState({ formError: null })
    }

    const formData = new FormData()
    formData.append("title", videoTitle)
    formData.append("description", videoDescription)
    formData.append("video", videoFile)
    formData.append("thumb", dataURLtoFile(thumbUrlData, "thumb"))
    this.onPostData(formData)
  }

  onPostData = (formData) => {
    fetch(`${process.env.REACT_APP_API_HOST}/video`, {
      method: "POST",
      body: formData,
    })
      .then((response) => response.json())
      .then(this.onFulfilledPost)
      .catch((e) => console.error(e))
  }

  onFulfilledPost = (data) => {
    if (data.errors && data.errors.length) {
      return this.setState({
        formError: data.errors
          .map((error) => `${error.param} - ${error.msg}`)
          .join("; "),
      })
    }

    const { history } = this.props

    return history.replace(`/video/${data.video.id}`, { ...data.video })
  }

  onFieldBlur = (value, fieldName) => {
    const fieldSettings = FORM_SCHEMA[fieldName]

    if (!fieldSettings) return null

    if (fieldSettings.isRequired) {
      const validationResult = isValidLength(
        value,
        fieldSettings.options,
        `Field value must be between ${fieldSettings.options.minLength} to ${fieldSettings.options.maxLength}`
      )

      switch (fieldName) {
        case "title": {
          this.setState({ title: { ...validationResult } })
          break
        }
        case "description": {
          this.setState({ description: { ...validationResult } })
          break
        }
        default: {
          return null
        }
      }
    }

    return null
  }

  onTitleChange = () => {
    this.setState({ title: { hasError: false, errorMessage: "" } })
  }

  onDescriptionChange = () => {
    this.setState({ description: { hasError: false, errorMessage: "" } })
  }

  onVideoChange = () => {
    this.setState({ video: { hasError: false, errorMessage: "" } })
  }

  render() {
    const { title, description, video, formError } = this.state

    return (
      <Container>
        <Row>
          <Column>
            <h2>Create new video</h2>
            <form className="upload-video__form" onSubmit={this.onSubmit}>
              <InputField
                {...FORM_SCHEMA.title}
                errorMessage={title.errorMessage}
                hasError={title.hasError}
                onChange={this.onTitleChange}
                onBlur={this.onFieldBlur}
              />
              <TextareaField
                {...FORM_SCHEMA.description}
                errorMessage={description.errorMessage}
                hasError={description.hasError}
                onChange={this.onDescriptionChange}
                onBlur={this.onFieldBlur}
              />
              <VideoUploader
                {...FORM_SCHEMA.video}
                onFileUpload={this.onVideoChange}
                errorMessage={video.errorMessage}
                hasError={video.hasError}
              />
              {formError && <FormErrorHint errorMessage={formError} />}
              <button className="button form_button--submit" type="submit">
                Submit
              </button>
            </form>
          </Column>
        </Row>
      </Container>
    )
  }
}

VideoFrom.propTypes = {
  history: ReactRouterPropTypes.history.isRequired,
}

export default VideoFrom
