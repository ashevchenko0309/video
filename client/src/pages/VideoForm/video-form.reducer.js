export const actions = {
  ON_CHANGE_TITLE: 'ON_CHANGE_TITLE',
  ON_CHANGE_DESCRIPTION: 'ON_CHANGE_DESCRIPTION',
  ON_CHANGE_VIDEO_UPLOAD: 'ON_CHANGE_VIDEO_UPLOAD',
  ON_VALIDATE_VIDEO_UPLOAD: 'ON_VALIDATE_VIDEO_UPLOAD'
}

export const initState = {
  title: {
    hasError: null,
    errorMessage: '',
  },
  description: {
    hasError: null,
    errorMessage: '',
  },
  video: {
    hasError: null,
    errorMessage: '',
  }
}

export function reducer(state, action) {
  switch (action.type) {
    case actions.ON_CHANGE_TITLE:
      return {
        ...state,
        title: { ...action.payload }
      }
    case actions.ON_CHANGE_DESCRIPTION:
      return {
        ...state,
        description: { ...action.payload }
      }
    case actions.ON_CHANGE_VIDEO_UPLOAD:
      return {
        ...state,
        video: { ...action.payload }
      }
    case actions.ON_VALIDATE_VIDEO_UPLOAD:
      return {
        ...state,
        video: { ...action.payload }
      }
    default:
      return state
  }
}