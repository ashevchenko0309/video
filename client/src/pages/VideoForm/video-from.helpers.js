const FORM_SCHEMA = {
  title: {
    name: "title",
    labelText: "Video name",
    placeholder: "Video name",
    isRequired: true,
    options: {
      minLength: 5,
      maxLength: 120
    }
  },
  description: {
    name: "description",
    labelText: "Video description",
    placeholder: "Video description",
    isRequired: true,
    options: {
      minLength: 5,
      maxLength: 255
    }
  },
  video: {
    name: "video",
    isRequired: true,
    accept: "video/mpeg,video/mp4"
  }
}

export const isValidLength = (value, options, errorMessage) => {
  if (value.length > options.maxLength) {
    return { hasError: true, errorMessage }
  }

  if (value.length < options.minLength) {
    return { hasError: true, errorMessage }
  }

  return { hasError: false, errorMessage: '' };
}

export default FORM_SCHEMA