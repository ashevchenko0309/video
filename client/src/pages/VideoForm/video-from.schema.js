const FORM_SCHEMA = {
  title: {
    name: "title",
    labelText: "Video name",
    placeholder: "Video name",
    isRequired: true,
    options: {
      minLength: 5,
      maxLength: 120,
    },
  },
  description: {
    name: "description",
    labelText: "Video description",
    placeholder: "Video description",
    isRequired: true,
    options: {
      minLength: 5,
      maxLength: 255,
    },
  },
  video: {
    name: "video",
    isRequired: true,
    accept: "video/mpeg,video/mp4",
  },
}

export default FORM_SCHEMA
