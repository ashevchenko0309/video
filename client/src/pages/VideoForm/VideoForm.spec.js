import React from "react"
import { createBrowserHistory } from 'history'

import VideoForm from "./VideoForm"
import isValidLength from "../../utils/validations"

const history = createBrowserHistory()

const DEFAULT_STATE = {
  formError: null,
  title: { hasError: false, errorMessage: "" },
  description: { hasError: false, errorMessage: "" },
  video: { hasError: false, errorMessage: "" },
}

const DEFAULT_ELEMENTS_VALUES = {
  elements: {
    title: { value: "Hello" },
    description: { value: "World" },
    video: { files: [{ value: "File" }] },
    thumb: {
      value:
        "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z/C/HgAGgwJ/lK3Q6wAAAABJRU5ErkJggg==",
    },
  },
}

const setUp = () => mount(<VideoForm history={history} />)

describe("VideoForm component", () => {
  let component
  let instance

  beforeEach(() => {
    component = setUp()
    instance = component.instance()
  })

  describe("Snapshots", () => {
    it("should render form (snapshot)", () => {
      expect(component.getElement()).toMatchSnapshot()
    })

    it("should render 2 inputs by default", () => {
      expect(component.find("input").length).toBe(2)
    })

    it("should render 4 inputs when upload file", () => {
      component.find(".form_field--upload").simulate("change", {
        target: { files: ["video"] },
      })

      setTimeout(() => {
        expect(component.find("input").length).toBe(4)
      }, 500)
    })
  })

  describe("Validation", () => {
    describe("Validation Errors", () => {
      it("should set error on submit empty fields", () => {
        expect(component.find(".form-error__hint").length).toBe(0)

        instance.onSubmit({
          preventDefault: () => { }, currentTarget: {
            elements: {
              title: { value: "" },
              description: { value: "" },
              video: { files: [{ value: "File" }] },
              thumb: { value: "" }
            }
          }
        })

        expect(component.state().formError).toBe(
          "Form have invalid value(s)"
        )
      })

      it("should show error if video field does not exist", () => {
        component.setState({ ...DEFAULT_STATE })

        instance.onSubmit({
          preventDefault: () => { },
          currentTarget: {
            elements: {
              ...DEFAULT_ELEMENTS_VALUES.elements,
              video: { files: [] },
            },
          },
        })

        expect(component.state().video).toStrictEqual({
          hasError: true,
          errorMessage: "Must have video file",
        })
      })

      it("should set form error on server error", () => {
        instance.onFulfilledPost({
          errors: [{ param: "title", msg: "invalid title" }],
        })
        expect(component.state().formError).toBe("title - invalid title")
      })

      it("should return error validation on less than minLength", () => {
        instance.onFulfilledPost({
          errors: [{ param: "title", msg: "invalid title" }],
        })
        expect(
          isValidLength("123", 5, 10)
        ).toStrictEqual({ hasError: true, errorMessage: "Value less than allowed length" })
      })

      it("should return error validation on grater than maxLength", () => {
        instance.onFulfilledPost({
          errors: [{ param: "title", msg: "invalid title" }],
        })
        expect(
          isValidLength("12345678910", 5, 10)
        ).toStrictEqual({ hasError: true, errorMessage: "Value greater than allowed length" })
      })
    })

    describe("Validation Success", () => {
      it("should hide error if formError has error", () => {
        component.setState({ ...DEFAULT_STATE, formError: "Some string" })

        instance.onSubmit({
          preventDefault: () => { },
          currentTarget: {
            elements: {
              title: { value: "Hello" },
              description: { value: "World" },
              video: { files: [{ value: "File" }] },
              thumb: {
                value:
                  "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mP8z/C/HgAGgwJ/lK3Q6wAAAABJRU5ErkJggg==",
              },
            },
          },
        })

        expect(component.state().formError).toBe(null)
      })

      it("should pass validation on title field", () => {
        instance.onFieldBlur("Hello", "title")
        expect(component.state().title).toStrictEqual({
          hasError: false,
          errorMessage: "",
        })
      })

      it("should pass validation on description field", () => {
        instance.onFieldBlur("World", "description")
        expect(component.state().description).toStrictEqual({
          hasError: false,
          errorMessage: "",
        })
      })

      it("should be silence on unknown field", () => {
        const result = instance.onFieldBlur("World", "unknown")
        expect(result).toBe(null)
      })

      it("should clear errors on change title", () => {
        component.setState({
          ...DEFAULT_STATE,
          title: { hasError: true, errorMessage: "Title invalid" },
        })

        instance.onTitleChange("Hello")
        expect(component.state().title).toStrictEqual({
          hasError: false,
          errorMessage: "",
        })
      })

      it("should clear errors on change description", () => {
        component.setState({
          ...DEFAULT_STATE,
          description: { hasError: true, errorMessage: "Title invalid" },
        })

        instance.onDescriptionChange("Hello")
        expect(component.state().description).toStrictEqual({
          hasError: false,
          errorMessage: "",
        })
      })

      it("should clear errors on change video", () => {
        component.setState({
          ...DEFAULT_STATE,
          video: { hasError: true, errorMessage: "Video invalid" },
        })

        instance.onVideoChange()
        expect(component.state().video).toStrictEqual({
          hasError: false,
          errorMessage: "",
        })
      })

      it("should return success validation", () => {
        instance.onFulfilledPost({
          errors: [{ param: "title", msg: "invalid title" }],
        })
        expect(
          isValidLength("12345", { minLength: 5, maxLength: 10 }, "Error")
        ).toStrictEqual({ hasError: false, errorMessage: "" })
      })
    })
  })
})
