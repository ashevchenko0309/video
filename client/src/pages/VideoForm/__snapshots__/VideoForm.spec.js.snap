// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoForm component Snapshots should render form (snapshot) 1`] = `
<VideoFrom
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
/>
`;
